godot-core/src/task/futures.rs

```diff
@@ -308,8 +335,162 @@ impl<C: WithBaseField, R: ParamTuple + Sync + Send> IntoFuture for &TypedSignal<
     }
 }
 
+/// Convert a value into a type that is [`Send`] at compile-time while the value might not be.
+///
+/// This allows to turn any implementor into a type that is `Send`, but requires to also implement [`DynamicSend`] as well.
+/// The later trait will verify if a value can actually be sent between threads at runtime.
+pub trait IntoMaybeSend: Sealed {
+    type Target: Send + DynamicSend<Target = Self>;
+
+    fn into_maybe_send(self) -> Self::Target;
+}
+
+/// Check that it was safe to send the value to the current thread.
+///
+/// # Safety
+/// The implementor has to guarantee that `assert_safe_send` panics if the value has been sent between threads while being `!Send`.
+pub unsafe trait DynamicSend: Send + Sealed {
+    type Target;
+
+    fn extract_if_safe(self) -> Option<Self::Target>;
+}
```

Comment:
```suggestion
/// Runtime-checked `Send` capability.
///
/// Implemented for types that need a static `Send` bound, but where it is determined at runtime whether sending a value was
/// actually safe. Only allows to extract the value if sending across threads is safe, thus fulfilling the `Send` supertrait.
///
/// # Safety
/// The implementor has to guarantee that `extract_if_safe` returns `None`, if the value has been sent between threads while being `!Send`.
///
/// To uphold the `Send` supertrait guarantees, no public API apart from `extract_if_safe` must exist that would give access to the inner value from another thread.
pub unsafe trait DynamicSend: Send + Sealed {
    type Target;

    fn extract_if_safe(self) -> Option<Self::Target>;
}
```



Added some elaboration, since

> DynamicSend: Send

is a bit unintuitive at first, because `DynamicSend` gives weaker guarantees than `Send`. Strictly speaking, a type is only `Send` in Rust if it can be safely transported across threads. So this trait asserts that the value is not accessible otherwise.